File: dead_reckoning_odom/src/DeadReckoning.cpp
// Header file for the class
#include "DeadReckoning.hpp"

// Namespace matches ROS package name
namespace dead_reckoning {

  // Constructor with global and private node handle arguments
  DeadReckoning::DeadReckoning(ros::NodeHandle n, ros::NodeHandle pn) {
    parent_frame = "odom";
    child_frame = "base_footprint";
    x = 0;
    y = 0;
    psi = 0;
    sample_time = 0.02;

    sub_twist = n.subscribe("diff_robot/twist", 1, &DeadReckoning::recvTwist, this);
    timer = n.createTimer(ros::Duration(sample_time), &DeadReckoning::timerCallback, this);
  }

  void DeadReckoning::timerCallback(const ros::TimerEvent& event) {
    // Integrate vehicle kinematics state space model one step with the latest speed and yaw rate data
    double v = twist_data.twist.linear.x;
    double pdot = twist_data.twist.angular.z;
    x += sample_time * v * cos(psi);
    y += sample_time * v * sin(psi);
    psi += sample_time * pdot;

    geometry_msgs::TransformStamped transform_msg;
    transform_msg.header.frame_id = parent_frame;
    transform_msg.header.stamp = event.current_real;
    transform_msg.child_frame_id = child_frame;

    // Copy dead reckoning estimate into transform message
    transform_msg.transform.translation.x = x;
    transform_msg.transform.translation.y = y;
    transform_msg.transform.translation.z = 0;

    // Populate quaternion corresponding to current yaw angle
    tf2::Quaternion q;
    q.setRPY(0, 0, psi);
    tf2::convert(q, transform_msg.transform.rotation);

    // Update TF transform with the latest dead reckoning iteration
    broadcaster.sendTransform(transform_msg);
  }

  void DeadReckoning::recvTwist(const geometry_msgs::TwistStampedConstPtr& msg) {
    twist_data = *msg;
  }
}

File: dead_reckoning_odom/scripts/dead_reckoning.py
#!/usr/bin/env python3
import math
import rospy
from geometry_msgs.msg import TwistStamped, TransformStamped
import tf
from tf2_ros import TransformBroadcaster


class DeadReckoning:
    def __init__(self):
        # Initialize variables
        self.twist_data = TwistStamped()
        self.x = 0
        self.y = 0
        self.psi = 0
        self.sample_time = 0.02

        # Load frame IDs from parameters
        self.parent_frame = 'odom'
        self.child_frame = 'base_footprint'

        self.sub_twist = rospy.Subscriber('diff_robot/twist', TwistStamped, self.recv_twist)
        self.timer = rospy.Timer(rospy.Duration(self.sample_time), self.timer_cb)
        self.broadcaster = TransformBroadcaster()

    def timer_cb(self, event):
        # TODO: Integrate vehicle navigation model one step with the latest speed and yaw rate data


        # TODO: Populate and broadcast a TF frame with the position and orientation from the model
        transform_msg = TransformStamped()

    def recv_twist(self, msg):
        self.twist_data = msg


if __name__ == '__main__':
    rospy.init_node('dead_reckoning')
    node_instance = DeadReckoning()
    rospy.spin()
